src/ast/ast_rewrite_same_clauses.c

```diff
@@ -0,0 +1,333 @@
+	
+	// build the replacement clause
+	cypher_astnode_t *new_clause = cypher_ast_create(false, pattern, &pattern,
+			1, range);
+
+	// replace original clause with the new one
+	cypher_ast_query_replace_clauses(root, new_clause, scope_start, scope_end);
+	
+	array_free(paths);
+}
+
+// compressing multiple consecutive MATCH clauses into a single MATCH clause
+// this function collects all patterns scattered across multiple MATCH clauses
+// and introduces them into a single MATCH clause
+static void replace_match_clause
+(
+	cypher_astnode_t *root,      // ast root
+	cypher_astnode_t **clauses,  // clause being replaced
+	int scope_start,             // beginning of scope
+	int scope_end                // ending of scope
+) {
+	uint count = array_len(clauses);
+
+	cypher_astnode_t *predicate = NULL;
+	struct cypher_input_range range = cypher_astnode_range(clauses[0]);
+	cypher_astnode_t **paths = array_new(cypher_astnode_t *, count);
+
+	// collect MATCH patterns and predicates
+	for (uint i = 0; i < count; i++) {
+		const cypher_astnode_t *pattern =
+			cypher_ast_match_get_pattern(clauses[i]);
+		uint npaths = cypher_ast_pattern_npaths(pattern);
+
+		for (uint j = 0; j < npaths; j++) {
+			const cypher_astnode_t *path =
+				cypher_ast_pattern_get_path(pattern, j);
+			array_append(paths, cypher_ast_clone(path));	
+		}
+
+		// combine MATCH predicates into a single predicate using AND connectors
+		cypher_astnode_t *new_predicate =
+			(cypher_astnode_t *)cypher_ast_match_get_predicate(clauses[i]);
+		if(new_predicate != NULL) {
+			new_predicate = cypher_ast_clone(new_predicate);
+			if(predicate == NULL) {
+				predicate = new_predicate;
+			} else {
+				// concat using AND
+				cypher_astnode_t *children[] = {predicate, new_predicate};
+				predicate = cypher_ast_binary_operator(CYPHER_OP_AND, predicate,
+						new_predicate, children, 2, range);
+			}
+		}
+		// done with current clause, free it
+		cypher_ast_free(clauses[i]);
+	}
+	
+	// build the replacement pattern
+	cypher_astnode_t *pattern =
+		cypher_ast_pattern(paths, array_len(paths), paths, array_len(paths),
+				range);
+	
+	// build the replacement clause
+	cypher_astnode_t *children[] = {pattern, predicate};
+	cypher_astnode_t *new_clause = cypher_ast_match(false, pattern, NULL, 0,
+			predicate, children, predicate == NULL ? 1 : 2, range);
+
+	// replace original clause with the new one
+	cypher_ast_query_replace_clauses(root, new_clause, scope_start, scope_end);
+	
+	array_free(paths);
+}
+
+// compressing multiple consecutive DELETE clauses into a single DELETE clause
+// this function collects all expressions scattered across multiple DELETE
+// clauses and introduces them into a single DELETE clause
+static void replace_delete_clause
+(
+	cypher_astnode_t *root,      // ast root
+	cypher_astnode_t **clauses,  // clause being replaced
+	int scope_start,             // beginning of scope
+	int scope_end                // ending of scope
+) {
```

Comment:
Let's try to remove duplicated entries in the final delete clause:
e.g.
```
DELETE n, a, n
DELETE a, n, z
```
Should be re-written as:
```
DELETE n, a, z
```